diff --git a/src/cairo-image-surface.c b/src/cairo-image-surface.c
index 21c9be4..70472d9 100644
--- a/src/cairo-image-surface.c
+++ b/src/cairo-image-surface.c
@@ -978,6 +978,21 @@ _cairo_image_surface_composite (cairo_operator_t   op,
 	if (status)
 	    goto CLEANUP_SURFACES;
 
+        if (mask_pattern->type == CAIRO_PATTERN_TYPE_SURFACE &&
+            (mask->width == 0 || mask->height == 0))
+       {
+           _cairo_pattern_release_surface (mask_pattern, &mask->base, &mask_attr);
+           mask = NULL;
+        }
+    }
+
+    if (src_pattern->type == CAIRO_PATTERN_TYPE_SURFACE &&
+        (src->width == 0 || src->height == 0))
+    {
+       /* nothing to do here */
+    }
+    else if (mask)
+    {
 	pixman_image_composite (_pixman_operator (op),
 				src->pixman_image,
 				mask->pixman_image,
